# 05. Mathematical Algorithms - Recursion
- Title: Mathematical Algorithms - Recursion in Python
- Date: Oct/06/2015, Tuesday - Current
- Author: Minwoo Bae (minubae.nyc@gmail.com)
- Reference: http://wphooper.com/teaching/2015-fall-308/python/Recursion.html

## Recursion
Recursion is the process of repeating items in a self-similar way. For instance, when the surfaces of two mirrors are exactly parallel with each other, the nested images that occur are a form of infinite recursion. The term has a variety of meanings specific to a variety of disciplines ranging from linguistics to logic. The most common application of recursion is in mathematics and computer science, in which it refers to a method of defining functions in which the function being defined is applied within its own definition. Specifically, this defines an infinite number of instances (function values), using a finite expression that for some instances may refer to other instances, but in such a way that no loop or infinite chain of references can occur. The term is also used more generally to describe a process of repeating objects in a self-similar way.

### Formal definitions 
In mathematics and computer science, a class of objects or methods exhibit recursive behavior when they can be defined by two properties:

- A simple base case (or cases)—a terminating scenario that does not use recursion to produce an answer
- A set of rules that reduce all other cases toward the base case

For example, the following is a recursive definition of a person's ancestors:

- One's parents are one's ancestors (base case).
- The ancestors of one's ancestors are also one's ancestors (recursion step).

In [4]:
# One thing to be wary of is that a function that calls itself can potentially go into an infinite loop. 
# For example the simplest possible recursive function is a total disaster because it does not terminate:
def dont_call_me():
    dont_call_me()

In [5]:
# To solve this problem, your function needs to make sure that every case eventually reaches a base case 
# in which the function does not call itself. Here is a simple recursive function that does terminate 
# (as long as n is a small integer.)
def call_me(n):
    if n<=0:
        print("Reached base case of n=",n)
    else:
        print("Function called with n=", n)
        call_me(n-1)
    print("Exiting call of function with n=", n)

In [6]:
# Below, I call this function. Can you figure out what is happening?
call_me(5) 

Function called with n= 5
Function called with n= 4
Function called with n= 3
Function called with n= 2
Function called with n= 1
Reached base case of n= 0
Exiting call of function with n= 0
Exiting call of function with n= 1
Exiting call of function with n= 2
Exiting call of function with n= 3
Exiting call of function with n= 4
Exiting call of function with n= 5


## Examples

##01) Factorial 
The factorial operation is usually defined inductively by $0!=1$ and $n!=n∗(n−1)!$ for integers $n≥1.$ This translates into the following function definition: 

In [7]:
def factorial(n):
    if n==0:
        return 1
    return n*factorial(n-1)

In [8]:
for n in range(7):
    print(n, "! = ", factorial(n), ".", sep="")

0! = 1.
1! = 1.
2! = 2.
3! = 6.
4! = 24.
5! = 120.
6! = 720.


##02) Pascal's triangle
Consider a polynomial $p(t)=a_{k}t^k+a_{k−1}t^{k−1}+\dots+a_{1}t+a_0.$ Observe that

$$(t+1)p(t)=a_{k}t^{k+1}+(a_k+a_{k−1})t^k+(a_{k−1}+a_{k−2})t^{k−1}+\dots+(a_0+a_1)t+a_0.$$
 
We can define a sequence of polynomials by $q0=1$ and inductively defining  q_{n+1}(t)=(t+1)q_{n}(t) for integers $n\geq0.$ Then each $q_{n}(t)$ is a polynomial of degree $n.$ We define a_{n,k} to be the coefficient of the  $t^k$ term of the polynomial $q_n(t).$ When arranged into a rectangular array, these numbers are known as Pascal's triangle.

<b>Problem:</b> Write a recursive function pascals_triangle(n,k) which returns  $a_{n,k}$ when provided an integer $n\geq0$ and an integer $k$  satisfying $0\leq k \leq n.$

To do this, first observe that the constant term of $q_n(t)$ is always one, as is the coefficient of $t^n$ in $q_{n}(t).$ That is, $q_{n,0} = 1$ and $q_{n,n} = 1$ for all integers $n\geq0.$<br>

Second, using the formula relating the coefficients for $(t+1)p(t)$ to the coefficients for $p(t)$ above, we see that $a_{n,k}=a_{n−1,k−1}+a_{n−1,k}$ when  $0<k<n.$<br>

The following function satisfies the requirements above.

In [9]:
def pascals_triangle(n,k):
    if k==0 or k==n:
        return 1
    return pascals_triangle(n-1,k-1)+pascals_triangle(n-1,k)

In [10]:
# The following prints out the coefficients of  (t+1)n(t+1)n  
# in successive lists:
for n in range(10):
    l=[]
    for k in range(n+1):
        l.append(pascals_triangle(n,k))
    print(l)

[1]
[1, 1]
[1, 2, 1]
[1, 3, 3, 1]
[1, 4, 6, 4, 1]
[1, 5, 10, 10, 5, 1]
[1, 6, 15, 20, 15, 6, 1]
[1, 7, 21, 35, 35, 21, 7, 1]
[1, 8, 28, 56, 70, 56, 28, 8, 1]
[1, 9, 36, 84, 126, 126, 84, 36, 9, 1]


The above is Pascal's triangle. You can read more about this in Wikipedia's article on <a href="https://en.wikipedia.org/wiki/Pascal%27s_triangle" target="_blank">Pascal's triangle</a>. Pascal's triangle is pretty neat. For instance, the locations of the odd integers in the triangle are related to the Sierpinski gasket (a fractal).<br>

<b>Remark:</b> The definition of Pascal's triangle is pretty slow. You can see this because if you think about the algorithm, to compute the 126 in the bottom row, we are adding 126 ones! There is a much faster version: <br>
$$a_{n,k}=\frac{n!}{k!(n−k)!}.$$


##03) Expansion of a number in base b
Let $b\geq2$ be an integer. Then any natural number $n$ has an expansion in base $b.$ This is the unique finite sequence of numbers $[a_0,a_1,\dots,a_k]$  so that
- each  $a_i \in \mathbb{Z}$ with $0\leq a_i<b,$
- $a_k\neq0,$ and
- $n=\displaystyle\sum_{i} a_{i} \cdot b^{i}.$

You are very familiar with base $10.$ The expansion of the number $123$ in base  $10$ is $a_0=3$, $a_1=2$ and $a3=1$ because <br>
$$123=3 \cdot 10^0+2 \cdot 10^1+1 \cdot 10^2.$$

We will explain how to solve the following problem: Give a function expansion(n,b) which returns the list $[a_0,a_1,\dots,a_k]$  representing the expansion of an integer $n\geq1$ in base $b,$ where  $b\geq2$ is an integer.

There are multiple ways to do this. One way is recursive. Suppose the expansion of $n$ is $[a_0,\dots,a_k]$ and $k>0.$ Then <br>
$$n=\displaystyle\sum_{i}^k a_{i} \cdot b^{i} = a_0 + b\bigg(\displaystyle\sum_{i}^{k-1} a_i \cdot b^i \bigg).$$

This shows two important things. First, $a_0$ is the remainder obtained when dividing  $n$ by $b$, i.e., $a0=n.$ Second, the sum in parenthesis above represents $\frac{n−a_0}{b}.$ The sum in these parentheses gives the expansion of this number. We need a base case for the recursion. We can use the fact that if $n<b$ then the expansion is just $[n].$

In [11]:
def expansion(n,b):
    if n<b:
        return [n]
    a0 = n%b
    lst=expansion((n-a0)//b, b)
    lst.insert(0,a0)
    return lst

In [12]:
expansion(123,10)

[3, 2, 1]

In [13]:
expansion(5,2)

[1, 0, 1]

##04) Towers of Hanoi

The <b>Tower of Hanoi</b> (also called the <b>Tower of Brahma</b> or <b>Lucas' Tower</b>, and sometimes pluralized) is a mathematical game or puzzle. It consists of three rods, and a number of disks of different sizes which can slide onto any rod. The puzzle starts with the disks in a neat stack in ascending order of size on one rod, the smallest at the top, thus making a conical shape.

The objective of the puzzle is to move the entire stack to another rod, obeying the following simple rules:

- Only one disk can be moved at a time.
- Each move consists of taking the upper disk from one of the stacks and placing it on top of another stack i.e. a disk can only be moved if it is the uppermost disk on a stack.
- No disk may be placed on top of a smaller disk.

With three disks, the puzzle can be solved in seven moves. The minimum number of moves required to solve a Tower of Hanoi puzzle is $2^n - 1,$ where $n$ is the number of disks.

##05) The Catalan Numbers
In combinatorial mathematics, the Catalan numbers form a sequence of natural numbers that occur in various counting problems, often involving recursively-defined objects. They are named after the Belgian mathematician Eugène Charles Catalan (1814–1894).

Using zero-based numbering, the nth Catalan number is given directly in terms of binomial coefficients by

$$C_n = \frac{1}{n+1}{2n\choose n} = \frac{(2n)!}{(n+1)!\,n!} = \prod\limits_{k=2}^{n}\frac{n+k}{k} \qquad\mbox{ for }n\ge 0.$$
 
 
The first Catalan numbers for n = 0, 1, 2, 3, … are

1, 1, 2, 5, 14, 42, 132, 429, 1430, 4862, 16796, 58786, 208012, 742900, 2674440, 9694845, 35357670, 129644790, 477638700, 1767263190, 6564120420, 24466267020, 91482563640, 343059613650, 1289904147324, 4861946401452, … 

They also satisfy: 

$$C_0=1 \space\space and \space\space C_{n+1}=\frac{2(2n+1)}{n+2}C_n \space\space for \space\space integers \space\space  n\geq 0.$$

Define a recursive function catalan(n) which takes as input an integer  $i\geq0$ and returns the Catalan number $C_i.$ (The Catalan numbers first appeared as an exercise in the page on Lists.)

<b>Remark:</b> Wikipedia has a discussion of this in their article on <a href="https://en.wikipedia.org/wiki/Catalan_number" target="_blank">the Catalan numbers</a>.

In [14]:
def catalan(n):
    if n == 0:
        return 1
    return catalan(n-1)*(2*(2*n+1)/(n+2))

In [15]:
catalan(10)

58786.0

##06) Multiplication using Addition
Write a recursive function multiply(m,n) which takes as input two integers  mm  and  $n,$ and returns their product $m∗n$ only using addition and subtraction. Hints:

$$m∗0=0,m∗n=m∗(n−1)+m \space\space and \space\space m∗n=m∗(n+1)−m.$$

<b>Remark:</b> Your function should work for all integers.

In [19]:
def multiply(m,n):
    try:
        if n==0:
            return 0
        elif n==1:
            return m
        elif n < 0:
            return  multiply(m, n+1) - m
        else:
            return multiply(m, n-1) + m
    except Exception as error:
        return error

In [20]:
multiply(3,2)

6

In [21]:
multiply(-3,10)

-30

##07) Greatest Common Divisor
Write a recursive function gcd(m,n) which returns the greatest common divisor of integers $m\geq0$ and $n\geq0$ (not both zero). Use the following observations:
- $gcd(m,n)=gcd(n,m)$ for all $m$ and $n.$
- $gcd(0,n)=n$ when $n>0.$
- $gcd(m,n)=gcd(n%m,m)$ whenever $0 < m \leq n.$


In [22]:
def gcd(m,n):
    if m==0:
        return n
    elif m>n:
        return gcd(m%n,n)
    return gcd(n%m, m)

In [23]:
gcd(10,6)

2

In [25]:
gcd(12,32)

4

##08) Iteration of a Function
Suppose $f:\mathbb{R} \to \mathbb{R}.$ Then, we define $f^{\circ k}$ to be $f$  applied $k \in \mathbb{N}$ times:

$$f^{\circ k}(x)=f \circ f \circ \dots \circ f^k(x)$$

for $x\in\mathbb{R}.$ Write a recursive function iterate(f,k,x) which takes as input a function $f:\mathbb{R} \to \mathbb{R}$, a natural number  $k,$ and a floating point number $x$ and returns the value of  $f^{\circ k}(x).$

In [26]:
def iterate(f,k,x):
    try:
        if k > 0:
            if k==1:
                return f(x)
            return iterate(f,k-1,f(x))
        else:
            return 'k is not greater than 1'
    except Exception as error:
        return error

In [27]:
g = lambda x: x**2
h = lambda x: 2*x*(1-x)
iterate(g,3,1.1)

2.143588810000001

In [29]:
iterate(h,100,0.25)

0.5

##09) The Cantor Set
In mathematics, the Cantor set is a set of points lying on a single line segment that has a number of remarkable and deep properties. It was discovered in 1874 by Henry John Stephen Smith and introduced by German mathematician Georg Cantor in 1883.

Through consideration of this set, Cantor and others helped lay the foundations of modern point-set topology. Although Cantor himself defined the set in a general, abstract way, the most common modern construction is the Cantor ternary set, built by removing the middle thirds of a line segment. Cantor himself only mentioned the ternary construction in passing, as an example of a more general idea, that of a perfect set that is nowhere dense.



<b>Remark:</b> Wikipedia has a discussion of this in their article on <a href="https://en.wikipedia.org/wiki/Cantor_set" target="_blank">the Cantor Set</a>.